File: src/openfl/events/Event.nim
import ./EventPhase

type
  Event* = ref object of RootObj
    bubbles*: bool
    cancelable*: bool
    currentTarget*: RootRef
    eventPhase*: EventPhase
    target*: RootRef
    `type`*: string
    isCanceled: bool
    isCanceledNow: bool
    preventDefaultFlag: bool

const
  ACTIVATE* = "activate"
  ADDED* = "added"
  ADDED_TO_STAGE* = "addedToStage"
  CANCEL* = "cancel"
  CHANGE* = "change"
  CLEAR* = "clear"
  CLOSING* = "closing"
  CLOSE* = "close"
  COMPLETE* = "complete"
  CONNECT* = "connect"
  CONTEXT3D_CREATE* = "context3DCreate"
  COPY* = "copy"
  CUT* = "cut"
  DEACTIVATE* = "deactivate"
  ENTER_FRAME* = "enterFrame"
  EXIT_FRAME* = "exitFrame"
  EXITING* = "exiting"
  FRAME_CONSTRUCTED* = "frameConstructed"
  FRAME_LABEL* = "frameLabel"
  FULLSCREEN* = "fullScreen"
  ID3* = "id3"
  INIT* = "init"
  MOUSE_LEAVE* = "mouseLeave"
  OPEN* = "open"
  PASTE* = "paste"
  REMOVED* = "removed"
  REMOVED_FROM_STAGE* = "removedFromStage"
  RENDER* = "render"
  RESIZE* = "resize"
  SCROLL* = "scroll"
  SELECT* = "select"
  SELECT_ALL* = "selectAll"
  SOUND_COMPLETE* = "soundComplete"
  TAB_CHILDREN_CHANGE* = "tabChildrenChange"
  TAB_ENABLED_CHANGE* = "tabEnabledChange"
  TAB_INDEX_CHANGE* = "tabIndexChange"
  TEXTURE_READY* = "textureReady"
  UNLOAD* = "unload"

proc newEvent*(typeStr: string, bubbles: bool = false, cancelable: bool = false): Event =
  Event(
    `type`: typeStr,
    bubbles: bubbles,
    cancelable: cancelable,
    eventPhase: EventPhase.AT_TARGET
  )

method clone*(self: Event): Event {.base.} =
  let event = newEvent(self.`type`, self.bubbles, self.cancelable)
  event.eventPhase = self.eventPhase
  event.target = self.target
  event.currentTarget = self.currentTarget
  return event

proc isDefaultPrevented*(self: Event): bool =
  self.preventDefaultFlag

proc preventDefault*(self: Event) =
  if self.cancelable:
    self.preventDefaultFlag = true

proc stopImmediatePropagation*(self: Event) =
  self.isCanceled = true
  self.isCanceledNow = true

proc stopPropagation*(self: Event) =
  self.isCanceled = true

method toString*(self: Event): string {.base.} =
  "[Event type=" & self.`type` & " bubbles=" & $self.bubbles & " cancelable=" & $self.cancelable & "]"

proc `$`*(self: Event): string =
  self.toString()

File: kfl/src/main/kotlin/openfl/events/Event.kt
package openfl.events

open class Event(
    val type: String,
    val bubbles: Boolean = false,
    val cancelable: Boolean = false
) {
    companion object {
        const val ACTIVATE = "activate"
        const val ADDED = "added"
        const val ADDED_TO_STAGE = "addedToStage"
        const val CANCEL = "cancel"
        const val CHANGE = "change"
        const val CLEAR = "clear"
        const val CLOSING = "closing"
        const val CLOSE = "close"
        const val COMPLETE = "complete"
        const val CONNECT = "connect"
        const val CONTEXT3D_CREATE = "context3DCreate"
        const val COPY = "copy"
        const val CUT = "cut"
        const val DEACTIVATE = "deactivate"
        const val ENTER_FRAME = "enterFrame"
        const val EXIT_FRAME = "exitFrame"
        const val EXITING = "exiting"
        const val FRAME_CONSTRUCTED = "frameConstructed"
        const val FRAME_LABEL = "frameLabel"
        const val FULLSCREEN = "fullScreen"
        const val ID3 = "id3"
        const val INIT = "init"
        const val MOUSE_LEAVE = "mouseLeave"
        const val OPEN = "open"
        const val PASTE = "paste"
        const val REMOVED = "removed"
        const val REMOVED_FROM_STAGE = "removedFromStage"
        const val RENDER = "render"
        const val RESIZE = "resize"
        const val SCROLL = "scroll"
        const val SELECT = "select"
        const val SELECT_ALL = "selectAll"
        const val SOUND_COMPLETE = "soundComplete"
        const val TAB_CHILDREN_CHANGE = "tabChildrenChange"
        const val TAB_ENABLED_CHANGE = "tabEnabledChange"
        const val TAB_INDEX_CHANGE = "tabIndexChange"
        const val TEXTURE_READY = "textureReady"
        const val UNLOAD = "unload"
    }

    var currentTarget: Any? = null
        internal set

    var target: Any? = null
        internal set

    var eventPhase: EventPhase = EventPhase.AT_TARGET
        internal set

    internal var isCanceled: Boolean = false
    internal var isCanceledNow: Boolean = false
    internal var isDefaultPrevented: Boolean = false

    open fun clone(): Event {
        val event = Event(type, bubbles, cancelable)
        event.eventPhase = eventPhase
        event.target = target
        event.currentTarget = currentTarget
        return event
    }

    fun isDefaultPrevented(): Boolean = isDefaultPrevented

    fun preventDefault() {
        if (cancelable) {
            isDefaultPrevented = true
        }
    }

    fun stopImmediatePropagation() {
        isCanceled = true
        isCanceledNow = true
    }

    fun stopPropagation() {
        isCanceled = true
    }

    override fun toString(): String {
        return "[Event type=$type bubbles=$bubbles cancelable=$cancelable]"
    }

    protected fun formatToString(className: String, vararg args: String): String {
        var output = "[$className"
        for (arg in args) {
            val value = when (arg) {
                "type" -> type
                "bubbles" -> bubbles
                "cancelable" -> cancelable
                "eventPhase" -> eventPhase
                else -> {
                    try {
                        val field = this::class.java.getDeclaredField(arg)
                        field.isAccessible = true
                        field.get(this)
                    } catch (e: Exception) {
                        null
                    }
                }
            }
            output += if (value is String) " $arg=\"$value\"" else " $arg=$value"
        }
        output += "]"
        return output
    }
}
